### PR TITLE
server/license: Integrate license throttle with SQL

### PR DESCRIPTION
**Pull request by spilchen**
Prior to this, the throttling code had been implemented but wasn’t integrated with SQL. This change completes that integration, applying throttling only to external SQL queries—internal SQL queries are exempt. If throttling is active, an error is returned and propagated to the caller.

For licenses requiring telemetry, the latest telemetry ping time is used in the throttling calculation by implementing the
license.TelemetryStatusReporter interface.

Most throttling functionality remains disabled, enabled only in unit tests or via environment variables. Full throttling will be enabled in CRDB-41758.

Epic: CRDB-39988
Closes: CRDB-39991
Release note: None

### CODE REVIEW

File: pkg/server/license/enforcer.go [hunk 1]
```diff
@@ -138,17 +145,34 @@ func (e *Enforcer) SetTelemetryStatusReporter(reporter TelemetryStatusReporter)
...
+	// is needed for testing that may be running this in isolation to the license
+	// ccl package.
+	e.RefreshForLicenseChange(ctx, LicTypeNone, time.Time{})
+
 	// Add a hook into the license setting so that we refresh our state whenever
-	// the license changes.
-	RegisterCallbackOnLicenseChange(ctx, st)
+	// the license changes. This will also update the state for the current
+	// license if not in test.
+	RegisterCallbackOnLicenseChange(ctx, st, e)
```
Comment (princejha95) [inline pkg/server/license/enforcer.go:171]: This method is also calling `RefreshForLicenseChange` and in line 166 also, we are calling the same method on the same enforcer. Any reason for this ?

**Review by fqazi — APPROVED** [review]
Nice work :lgtm:

Just very minor suggestions

> Reviewed 3 of 8 files at r1, 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130263#-:-O6TKkf855SIbZwX4wjx:b-oshxtz)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen and @xinhaoz)

___
*[`pkg/server/license/enforcer_test.go` line 273 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130263#-O6TL3A44MpBUrODyAyC:-O6TL3A5546tgCjyO8YP:bvvjdec) ([raw file](https://github.com/cockroachdb/cockroach/blob/eaa6bec4e3fb614f3b15481440963f9de32c1f8e/pkg/server/license/enforcer_test.go#L273)):*
> ```Go
> 					_, err = tx.Exec("SELECT 1")
> 					if err != nil {
> 						return err
> ```

Nit: Also think there is a risk of a hang if something goes wrong here. I don't think we need real concurrency here, if you want to simplify the test design.
___
*[`pkg/server/license/enforcer_test.go` line 308 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130263#-O6TKr257lzgfcY1vW1B:-O6TKr257lzgfcY1vW1C:bvget9u) ([raw file](https://github.com/cockroachdb/cockroach/blob/eaa6bec4e3fb614f3b15481440963f9de32c1f8e/pkg/server/license/enforcer_test.go#L308)):*
> ```Go
> 
> 			// Unblock the background Go routines and wait for them to finish.
> 			close(unblockSignal)
> ```

Nit: Lets do a defer on this to be safer

**Review by spilchen — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130263)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @fqazi and @xinhaoz)

___
*[`pkg/server/license/enforcer_test.go` line 273 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130263#-O6TL3A44MpBUrODyAyC:-O6W_wqw3lEnKPga9Ayh:b-cqmspq) ([raw file](https://github.com/cockroachdb/cockroach/blob/eaa6bec4e3fb614f3b15481440963f9de32c1f8e/pkg/server/license/enforcer_test.go#L273)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Nit: Also think there is a risk of a hang if something goes wrong here. I don't think we need real concurrency here, if you want to simplify the test design.
</blockquote></details>

Good suggestion. I noticed the test would lock up during implementation, so I think you're right—it could become an issue later. I added a testing knob and removed the concurrency. Now, if the test exceeds the transaction limit, I set the max open transactions limit to 0.
___
*[`pkg/server/license/enforcer_test.go` line 308 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130263#-O6TKr257lzgfcY1vW1B:-O6WajcW6LOTXowH5phP:b-nutyge) ([raw file](https://github.com/cockroachdb/cockroach/blob/eaa6bec4e3fb614f3b15481440963f9de32c1f8e/pkg/server/license/enforcer_test.go#L308)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Nit: Lets do a defer on this to be safer
</blockquote></details>

I removed the concurrency, so I no longer do this.

**Comment by spilchen** [comment]
TFTR @fqazi 

bors r=fqazi

**Comment by spilchen** [comment]
blathers backport 23.1 23.2 24.1 24.2

**Pull request merged**